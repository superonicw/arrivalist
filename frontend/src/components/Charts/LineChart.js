import React, { useEffect, useState, useMemo } from 'react'
import { LineChart as D3LineChart } from 'react-d3-components'
import { find, flattenDeep, get, maxBy, minBy, values } from 'lodash'
import moment from 'moment'
import * as d3 from 'd3'
import Tooltip from './Tooltip'
import { STATES } from 'config/constants'

const LineChart = ({ trips }) => {
  const [viewport, setViewport] = useState({ width: 0, height: 0 })

  useEffect(() => {
    handleResize()
    window.addEventListener('resize', handleResize)

    return () => {
      window.removeEventListener('resize', handleResize)
    }
  }, [])

  function handleResize() {
    const wrapper = document.querySelector('.linechart')

    if (!wrapper) {
      return
    }

    const { width, height } = wrapper.getBoundingClientRect()

    setViewport({ width, height })
  }

  const data = useMemo(() => {
    if (trips.length === 0) {
      return null
    }

    const res = trips.reduce((acc, elem) => {
      const { home_state, trip_date, trip_count } = elem

      if (!acc[home_state]) {
        acc[home_state] = {
          key: home_state,
          values: {},
        }
      }

      if (!acc[home_state]['values'][trip_date]) {
        const date = Date.parse(trip_date)

        acc[home_state]['values'][trip_date] = {
          state: get(find(STATES, { abbr: home_state }), 'name'),
          x: date,
          y: 0,
        }
      }

      acc[home_state]['values'][trip_date]['y'] += trip_count

      return acc
    }, {})

    const lines = values(res).map(elem => ({
      ...elem,
      values: values(elem.values),
    }))

    const allPoints = flattenDeep(lines.map(line => line.values))
    const minDate = minBy(allPoints, 'x').x
    const maxDate = maxBy(allPoints, 'x').x

    return { lines, minDate, maxDate }
  }, [trips])

  const margin = { top: 30, bottom: 80, left: 50, right: 30 }

  function renderChart() {
    if (!data) {
      return null
    }

    const { lines, minDate, maxDate } = data

    const xScale = d3.time.scale().domain([minDate, maxDate]).range([0, 400])
    const xInterval = Math.round(
      moment(maxDate).diff(moment(minDate), 'days') / 5,
    )

    return (
      <D3LineChart
        data={lines}
        width={viewport.width}
        height={viewport.height}
        margin={margin}
        tooltipHtml={(_, props) => <Tooltip {...props} />}
        xAxis={{
          tickValues: xScale.ticks(d3.time.day, xInterval),
          tickFormat: d3.time.format('%b %e'),
          tickPadding: 20,
          innerTickSize: 5,
          orient: 'right',
        }}
        yAxis={{
          tickFormat: d3.format('s'),
          tickPadding: 20,
          innerTickSize: 5,
        }}
        shapeColor="red"
        stroke="red"
        label={line => line.key}
        colorScale={abbr => get(find(STATES, { abbr }), 'color')}
      />
    )
  }

  return <div className="linechart">{renderChart()}</div>
}

export default LineChart
