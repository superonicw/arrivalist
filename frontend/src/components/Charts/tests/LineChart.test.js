import React from 'react'
import { mount, shallow } from 'enzyme'
import { TripMock } from 'test/mocks'
import LineChart from '../LineChart'

const initialProps = {
  trips: TripMock,
}

describe('LineChart test', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<LineChart {...initialProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders without chart', () => {
    const wrapper = mount(<LineChart {...initialProps} />)
    wrapper.setProps({ trips: [] })

    expect(wrapper).toMatchSnapshot()

    wrapper.unmount()
  })
})
