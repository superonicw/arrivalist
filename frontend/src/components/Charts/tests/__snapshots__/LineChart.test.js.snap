// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart test renders correctly 1`] = `
<div
  className="linechart"
>
  <LineChart
    barPadding={0.5}
    colorScale={[Function]}
    customPointColor="blue"
    customPointShape="circle"
    data={
      Array [
        Object {
          "key": "AK",
          "values": Array [
            Object {
              "state": "Alaska",
              "x": 1598918400000,
              "y": 1657,
            },
            Object {
              "state": "Alaska",
              "x": 1599004800000,
              "y": 1472,
            },
            Object {
              "state": "Alaska",
              "x": 1599091200000,
              "y": 2400,
            },
            Object {
              "state": "Alaska",
              "x": 1599177600000,
              "y": 7212,
            },
            Object {
              "state": "Alaska",
              "x": 1599264000000,
              "y": 4976,
            },
            Object {
              "state": "Alaska",
              "x": 1599350400000,
              "y": 2166,
            },
            Object {
              "state": "Alaska",
              "x": 1599436800000,
              "y": 1099,
            },
            Object {
              "state": "Alaska",
              "x": 1599523200000,
              "y": 1317,
            },
            Object {
              "state": "Alaska",
              "x": 1599609600000,
              "y": 1678,
            },
          ],
        },
        Object {
          "key": "AL",
          "values": Array [
            Object {
              "state": "Alabama",
              "x": 1598918400000,
              "y": 22010,
            },
            Object {
              "state": "Alabama",
              "x": 1599004800000,
              "y": 23263,
            },
            Object {
              "state": "Alabama",
              "x": 1599091200000,
              "y": 34660,
            },
            Object {
              "state": "Alabama",
              "x": 1599177600000,
              "y": 71158,
            },
            Object {
              "state": "Alabama",
              "x": 1599264000000,
              "y": 54976,
            },
            Object {
              "state": "Alabama",
              "x": 1599350400000,
              "y": 27697,
            },
            Object {
              "state": "Alabama",
              "x": 1599436800000,
              "y": 20603,
            },
            Object {
              "state": "Alabama",
              "x": 1599523200000,
              "y": 23859,
            },
            Object {
              "state": "Alabama",
              "x": 1599609600000,
              "y": 23616,
            },
          ],
        },
      ]
    }
    defined={[Function]}
    height={0}
    interpolate="linear"
    label={[Function]}
    lineStructureClassName="dot"
    margin={
      Object {
        "bottom": 80,
        "left": 50,
        "right": 30,
        "top": 30,
      }
    }
    shape="circle"
    shapeColor="red"
    showCustomLine={false}
    stroke="red"
    tooltipContained={false}
    tooltipHtml={[Function]}
    tooltipMode="mouse"
    tooltipOffset={
      Object {
        "left": 0,
        "top": -35,
      }
    }
    values={[Function]}
    width={0}
    x={[Function]}
    xAxis={
      Object {
        "innerTickSize": 5,
        "orient": "right",
        "tickFormat": [Function],
        "tickPadding": 20,
        "tickValues": Array [
          2020-09-01T00:00:00.000Z,
          2020-09-03T00:00:00.000Z,
          2020-09-05T00:00:00.000Z,
          2020-09-07T00:00:00.000Z,
          2020-09-09T00:00:00.000Z,
        ],
      }
    }
    xScale={null}
    y={[Function]}
    y0={[Function]}
    yAxis={
      Object {
        "innerTickSize": 5,
        "tickFormat": [Function],
        "tickPadding": 20,
      }
    }
    yScale={null}
  />
</div>
`;

exports[`LineChart test renders without chart 1`] = `
<LineChart
  trips={Array []}
>
  <div
    className="linechart"
  />
</LineChart>
`;
